libs/execution_context/ebpf_maps.c

```diff
@@ -1179,6 +1186,203 @@ _update_stack_map_entry(
+        ebpf_ring_buffer_map_async_query_result_t* async_query_result = context->async_query_result;
+        ebpf_ring_buffer_query(
+            (ebpf_ring_buffer_t*)map->data, &async_query_result->consumer, &async_query_result->producer);
+        ebpf_list_remove_entry(&context->entry);
+        ebpf_async_complete(context->async_context, sizeof(*async_query_result), EBPF_SUCCESS);
+        ebpf_free(context);
+        context = NULL;
+    }
+}
+
+static void
+_delete_ring_buffer_map(_In_ ebpf_core_map_t* map)
+{
+    EBPF_LOG_ENTRY();
+    // Free the ring buffer.
+    ebpf_ring_buffer_destroy((ebpf_ring_buffer_t*)map->data);
+
+    ebpf_core_ring_buffer_map_t* ring_buffer_map = EBPF_FROM_FIELD(ebpf_core_ring_buffer_map_t, core_map, map);
+    // Snap the async context list.
+    ebpf_list_entry_t temp_list;
+    ebpf_list_initialize(&temp_list);
+    ebpf_lock_state_t state = ebpf_lock_lock(&map->lock);
+    ebpf_list_entry_t* first_entry = ring_buffer_map->async_contexts.Flink;
+    if (!ebpf_list_is_empty(&ring_buffer_map->async_contexts)) {
+        ebpf_list_remove_entry(&ring_buffer_map->async_contexts);
+        ebpf_list_append_tail_list(&temp_list, first_entry);
+    }
+    ebpf_lock_unlock(&map->lock, state);
+    // Cancel all pending async query operations.
+    for (ebpf_list_entry_t* temp_entry = temp_list.Flink; temp_entry != &temp_list; temp_entry = temp_entry->Flink) {
+        ebpf_core_ring_buffer_map_async_query_context_t* context =
+            EBPF_FROM_FIELD(ebpf_core_ring_buffer_map_async_query_context_t, entry, temp_entry);
+        ebpf_async_complete(context->async_context, 0, EBPF_CANCELLED);
+    }
+    ebpf_epoch_free(ring_buffer_map);
+}
+
+static ebpf_core_map_t*
+_create_ring_buffer_map(_In_ const ebpf_map_definition_in_memory_t* map_definition)
+{
+    ebpf_result_t result;
+    ebpf_core_ring_buffer_map_t* ring_buffer_map = NULL;
+    ebpf_ring_buffer_t* ring_buffer = NULL;
+    ebpf_core_map_t* core_map = NULL;
+
+    EBPF_LOG_ENTRY();
+
+    // allocate
+    ring_buffer_map = ebpf_epoch_allocate(sizeof(ebpf_core_ring_buffer_map_t));
+    if (ring_buffer_map == NULL) {
+        result = EBPF_NO_MEMORY;
+        goto Exit;
+    }
+    memset(ring_buffer_map, 0, sizeof(ebpf_core_ring_buffer_map_t));
+
+    ring_buffer_map->core_map.ebpf_map_definition = *map_definition;
+    result =
+        ebpf_ring_buffer_create((ebpf_ring_buffer_t**)&ring_buffer_map->core_map.data, map_definition->max_entries);
+    if (result != EBPF_SUCCESS)
+        goto Exit;
+    ring_buffer = (ebpf_ring_buffer_t*)ring_buffer_map->core_map.data;
+
+    ebpf_list_initialize(&ring_buffer_map->async_contexts);
+
+    core_map = &ring_buffer_map->core_map;
+    ring_buffer = NULL;
+    ring_buffer_map = NULL;
+
+Exit:
+    if (ring_buffer != NULL)
+        ebpf_ring_buffer_destroy(ring_buffer);
+
+    if (ring_buffer_map != NULL)
+        ebpf_epoch_free(ring_buffer_map);
```

Comment:
Ok, but would be safer and more consistent with standard apis to make these accept nullptr, like free() does.